python/ray/train/_internal/session.py

```diff
@@ -556,6 +559,16 @@ def new_report(
 
         metrics = self._auto_fill_metrics(metrics)
 
+        # Set additional user metadata from the Trainer.
+        if persisted_checkpoint and self.metadata:
+            user_metadata = persisted_checkpoint.get_metadata()
+            for k, v in self.metadata.items():
+                # Update keys not already set by the user. This gives user-set keys
+                # precedence over keys set at the Trainer level.
+                if k not in user_metadata:
+                    user_metadata[k] = v
+            persisted_checkpoint.set_metadata(user_metadata)
+
```

Comment:
We will be setting the metadata many times here, once for each worker. Can we guard this with a rank check (e.g. only report from rank 0 worker) and only set metadata once?

The other caveat here is that other trainers (xgb, lgbm, sklearn) don't have Train workers calling `train.report`, so we can't access `train.get_context().get_world_rank()` there.

Reply:
Hmm, I feel like this would be more brittle given we are now supporting not reporting from rank 0. I don't think the performance impact here is measurable.

Reply:
Ok, not to big of a deal